cmd/vmars/draw.go

```diff
@@ -4,73 +4,203 @@ import (
 	"fmt"
 	"image"
 	"image/color"
+	"math"
+	"sort"
 
 	"github.com/bobertlo/gmars"
 	"github.com/hajimehoshi/ebiten/v2"
+	"github.com/hajimehoshi/ebiten/v2/colorm"
 	"github.com/hajimehoshi/ebiten/v2/text/v2"
+	"golang.org/x/image/colornames"
 )
 
-func (g *Game) Draw(screen *ebiten.Image) {
-	warriorColors := make([]ebiten.ColorScale, 3)
-	warriorColors[0].Scale(1, 1, 1, 1)
-	warriorColors[1].Scale(1, 1, 0, 1)
-	warriorColors[2].Scale(0, 1, 1, 1)
+const (
+	SpriteFirst = iota
+	SpriteHead
+	SpriteHeadActive
+	SpriteExecuted
+	SpriteCode
+	SpriteData
+	SpriteIncr
+	SpriteDecr
+	SpriteRead
+	SpriteDead
+	SpriteLast
+)
 
-	warriorQueueColors := make([]ebiten.ColorScale, 2)
-	warriorQueueColors[0].Scale(0.5, 0.5, 0.15, 1)
-	warriorQueueColors[1].Scale(0.15, 0.5, 0.5, 1)
+const screenXTileCount = screenWidth / tileSize
+const minimumHeadTransparency = 0.5
+const offsetHeadSprites = tileSize / 2.0
 
-	execColor := ebiten.ColorScale{}
-	execColor.Scale(1, 0.25, 0.25, 1)
+// cacheSprites buffers specific tiles from the tile image in different hue
+// variants.  This helps performance a lot, because we don't need to shuffle so
+// much memory on ever Update anymore.
+func (g *Game) cacheSprites() {
+	g.spriteCache = make([][]*ebiten.Image, SpriteLast)
+	sprnum := SpriteFirst + 1
 
-	w := tilesImage.Bounds().Dx()
-	tileXCount := w / tileSize
+	tileXCount := tilesImage.Bounds().Dx() / tileSize
+	for y := 0; y < tileXCount; y++ {
+		for x := 0; x < tileXCount; x++ {
+			sx := x * tileSize
+			sy := y * tileSize
+			rect := image.Rect(sx, sy, sx+tileSize, sy+tileSize)
 
-	const xCount = screenWidth / tileSize
+			subImage := tilesImage.SubImage(rect).(*ebiten.Image)
 
-	for i := 0; i < int(g.sim.CoreSize()); i++ {
-		state, color := g.rec.GetMemState(gmars.Address(i))
+			g.spriteCache[sprnum] = make([]*ebiten.Image, len(g.hues))
+			for i, hue := range g.hues {
+				var c colorm.ColorM
+				c.ChangeHSV(hue/180.0*math.Pi, 1.0, 1.0)
```

Comment:
I couldn't figure out any other way to vary hue. The color scale thing is RGB, so not that useful.